File: app/src/main/java/com/icerrate/popularmovies/view/movies/MoviesCatalogAdapter.java
package com.icerrate.popularmovies.view.movies;

import android.content.Context;
import android.support.v7.widget.RecyclerView;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ImageView;

import com.icerrate.popularmovies.R;
import com.icerrate.popularmovies.data.model.Movie;
import com.icerrate.popularmovies.view.common.GlideApp;
import com.icerrate.popularmovies.view.common.LoadMoreBaseAdapter;

import java.util.ArrayList;

/**
 * @author Ivan Cerrate.
 */

public class MoviesCatalogAdapter extends LoadMoreBaseAdapter<Movie> {

    private OnItemClickListener onItemClickListener;

    private int columns;

    public MoviesCatalogAdapter(OnItemClickListener onItemClickListener, int columns) {
        this(new ArrayList<Movie>(), onItemClickListener, columns);
    }

    public MoviesCatalogAdapter(ArrayList<Movie> movies, OnItemClickListener onItemClickListener, int columns) {
        this.data = movies;
        this.onItemClickListener = onItemClickListener;
        this.columns = columns;
    }

    @Override
    public RecyclerView.ViewHolder onCreateDataViewHolder(ViewGroup parent, int viewType) {
        View layoutView = LayoutInflater.from(parent.getContext()).inflate(R.layout.item_movies_catalog, parent, false);
        layoutView.getLayoutParams().height = (int) (parent.getWidth() / columns *
                1.5f);
        return new MovieViewHolder(layoutView, onItemClickListener);
    }

    @Override
    public void onBindDataViewHolder(RecyclerView.ViewHolder holder, int position) {
        if (holder instanceof MovieViewHolder) {
            MovieViewHolder movieViewHolder = (MovieViewHolder) holder;
            Movie movie = data.get(position);
            Context context = holder.itemView.getContext();
            String url = movie.getPosterUrl("w500");

            GlideApp.with(context)
                    .load(url)
                    .placeholder(context.getResources().getDrawable(R.drawable.poster_placeholder))
                    .into(movieViewHolder.posterImageView);

            movieViewHolder.itemView.setTag(movie);
        }
    }

    public void addItems(ArrayList<Movie> items) {
        this.data.addAll(items);
        notifyDataSetChanged();
    }

    public void resetItems() {
        this.data.clear();
        notifyDataSetChanged();
    }

    public class MovieViewHolder extends RecyclerView.ViewHolder implements View.OnClickListener {

        private ImageView posterImageView;

        private OnItemClickListener onItemClickListener;

        public MovieViewHolder(View itemView, OnItemClickListener onItemClickListener) {
            super(itemView);
            posterImageView = itemView.findViewById(R.id.poster);
            this.onItemClickListener = onItemClickListener;
            itemView.setOnClickListener(this);
        }

        @Override
        public void onClick(View view) {
            if (onItemClickListener != null) {
                onItemClickListener.onItemClick(view);
            }
        }
    }

    public interface OnItemClickListener {

        void onItemClick(View view);
    }
}

File: app/src/main/java/com/icerrate/popularmovies/view/movies/catalog/MoviesCatalogAdapter.kt
package com.icerrate.popularmovies.view.movies.catalog

import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import androidx.core.content.ContextCompat
import androidx.recyclerview.widget.RecyclerView
import com.bumptech.glide.Glide
import com.icerrate.popularmovies.R
import com.icerrate.popularmovies.data.model.Movie
import com.icerrate.popularmovies.databinding.ItemMoviesCatalogBinding
import com.icerrate.popularmovies.view.common.LoadMoreBaseAdapter

/**
 * @author Ivan Cerrate.
 */
class MoviesCatalogAdapter @JvmOverloads constructor(
    movies: ArrayList<Movie> = arrayListOf(),
    private val onItemClickListener: OnItemClickListener?,
    private val columns: Int
) : LoadMoreBaseAdapter<Movie>() {

    init {
        this.data = movies.toMutableList()
    }

    override fun onCreateDataViewHolder(parent: ViewGroup, viewType: Int): RecyclerView.ViewHolder {
        val binding = ItemMoviesCatalogBinding.inflate(LayoutInflater.from(parent.context), parent, false)
        binding.root.layoutParams.height = (parent.width / columns * RATIO).toInt()
        return MovieViewHolder(binding, onItemClickListener)
    }

    override fun onBindDataViewHolder(holder: RecyclerView.ViewHolder, position: Int) {
        if (holder is MovieViewHolder) {
            val movie = data[position] as Movie
            holder.bind(movie)
        }
    }

    fun addItems(items: List<Movie>) {
        this.data.addAll(items)
        notifyDataSetChanged()
    }

    fun resetItems() {
        this.data.clear()
        notifyDataSetChanged()
    }

    inner class MovieViewHolder(
        private val binding: ItemMoviesCatalogBinding,
        private val onItemClickListener: OnItemClickListener?
    ) : RecyclerView.ViewHolder(binding.root), View.OnClickListener {

        init {
            itemView.setOnClickListener(this)
        }

        fun bind(movie: Movie) {
            val context = binding.root.context
            val url = movie.getPosterUrl("w500")

            Glide.with(context)
                .load(url)
                .placeholder(ContextCompat.getDrawable(context, R.drawable.poster_placeholder))
                .error(ContextCompat.getDrawable(context, R.drawable.poster_placeholder))
                .into(binding.poster)

            binding.root.tag = movie
        }

        override fun onClick(view: View) {
            onItemClickListener?.onItemClick(view)
        }
    }

    interface OnItemClickListener {
        fun onItemClick(view: View)
    }

    companion object {
        private const val RATIO = 1.5f
    }
}